olive/passes/onnx/quantization.py

```diff
@@ -506,3 +506,51 @@ def _default_config(accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigPa
         # external data config
         config.update(get_external_data_config())
         return config
+
+
+class OnnxMatMul4Quantizer(Pass):
+    @staticmethod
+    def _default_config(accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        config = {
+            "block_size": PassConfigParam(
+                type_=int,
+                default_value=32,
+                description="Block size for quantization. Default value is 32.",
+            ),
+            "is_symmetric": PassConfigParam(
+                type_=bool,
+                default_value=True,
+                description="Symmetric quantization. Default value is True.",
+            ),
+            "nodes_to_exclude": PassConfigParam(
+                type_=list,
+                default_value=[],
+                description="List of node names to exclude from quantization.",
+            ),
+        }
+        config.update(get_external_data_config())
+        return config
+
+    def _run_for_config(
+        self, model: ONNXModel, data_root: str, config: Dict[str, Any], output_model_path: str
+    ) -> ONNXModel:
+        from onnxruntime import __version__ as OrtVersion
+
+        if version.parse(OrtVersion) < version.parse("1.17.0"):
+            raise OlivePassError("OnnxLlamaMatMulWeight4Quantizer is only supported in onnxruntime >= 1.17.0")
+
+        from onnxruntime.quantization.matmul_4bits_quantizer import MatMul4BitsQuantizer
+
+        quant = MatMul4BitsQuantizer(
+            model.load_model(), config["block_size"], config["is_symmetric"], config["nodes_to_exclude"]
+        )
+        quant.process()
+
+        new_tmp_dir = tempfile.TemporaryDirectory(prefix="olive_tmp")
```

Comment:
Is this tmp_dir necessary? 
Since the quantizer edits the model in place, I think there are two options:
```
onnx_model = model.load_model()
quant = ...
return model_proto_to_olive_model(onnx_model, ...)
```
or 
```
quant =  MatMul4BitsQuantizer(model.load_model(), ...)
...
return model_proto_to_olive_model(quant.model.model, ....)
```
Both would work but I think the first option makes it clearer without assuming anything about the structure of `quant`.

Reply:
![image](https://github.com/microsoft/Olive/assets/13343117/6238b848-a9be-4d13-a6b6-e99168541bf8)
The ONNXModel from ort quantization save the model with:
1. topological_srt
2. extra arguemnts like convert_attribute as True.
To align with the original implementation, I prefer to save with the `save_model_to_file` firstly, then save with Olive's method.

Reply:
Added TODO. We can refine this part later.
![image](https://github.com/microsoft/Olive/assets/13343117/aab82930-18fc-4036-956b-bd33d40d2cb2)

Reply:
Yes, I think we can expose the `convert_attribute` in our save helper function. 
In this case, `saving -> loading -> saving` undoes the original `convert_attribute` setting since it is only used during saving. After the model is loaded, the external data information is wiped out. So the original save to temp file is redundant. 

The only thing that a direct save doesn't do is the `topological sort`. This part, we can do ourselves like in https://github.com/microsoft/Olive/blob/main/olive/passes/onnx/transformer_optimization.py#L207 

But this is not a blocker. Just thought its cleaner if we don't deal with temp files when not needed.

Reply:
Agree. Let us finish it in following PR.